docs/designs/supporting_pytorch.md

```diff
@@ -0,0 +1,297 @@
+To go further, we can use `model.parameters()` and `model.zero_grad()` (defined by PyTorch for `nn.Module`) to make these steps more concise, and there will be no errors of forgetting some parameters, especially when we build a complex model:
+```python
+with torch.no_grad():
+    for param in model.parameters(): param -= param.grad * lr
+    model.zero_grad()
+```
+
+
+## Work with PS_client
+This [document](https://github.com/sql-machine-learning/elasticdl/blob/develop/docs/designs/parameter_server.md) describes the design of a distributed parameter server for ElasticDL.
+```python
+class PSClient(object):
+    def __init__(self, ps_stubs):
+        self.ps_stubs = ps_stubs
+        self.ps_num = len(self.ps_stubs)
+        self.parameter_to_ps = {}
+        self.ps_to_parameter = {}
+
+    def pull_embedding_vectors(self, layer_name, embedding_ids):
+        return new_embeddings
+
+    def partition_dense_parameters(self, param_names):
+        # Partition dense parameters to PS
+        # ps_id = string_to_id(param_name)
+        
+    def push_dense_parameters(self, parameters, ps_id, version):
+        # Push dense parameters to PS
+        # Args:parameters: a list of Tensors
+        #     ps_id: PS id
+        #     version: model version
+
+    def push_gradients(self, grads, edl_grads, learning_rate, model_versions,):
+        # Push gradients to PS. There two kinds of gradients:
+        # - gradients of normal layers
+        # - sparse gradients of ElasticDL embedding layers
+        # 1. handle grads
+        # 2. handle sparse grads of elasticdl embedding layers
+        # Sum up the values of the duplicated indices in the
+        # gradients. It can reduce the gradient payload of the
+        # dense embedding.
+        # 3. push gradients to PS
+```
+
+### Model Parameter Access from Worker
+Each PS pod has a RPC servicer to provide RPC services. Workers use RPC services to pull model parameters. `pull_variable` service is to pull all non-embedding parameters. `pull_embedding_vector` service is to pull embedding vectors specified by an embedding layer name and a list of discrete IDs.
+```python
+service PServer{
+    rpc pull_variable(PullModelRequest) returns (PullModelResponse);
+    rpc pull_embedding_vector(PullEmbeddingVectorRequest) returns (Tensor);
+}
+```
+### Model Parameter Update
+A worker computes gradients in each training iteration, which contain gradients for non-embedding parameters and some embedding vectors if applicable. The worker partitions these gradients using their corresponding parameter names or discrete IDs for embedding vectors. Then the worker sends gradient partitions to their corresponding PS pods by RPC calls `push_gradient`.
+```python
+service PServer{
+    rpc push_gradient(PushGradientRequest) returns (PushGradientResponse);
+}
+```
+When a PS pod receives gradients in `push_gradient`, it uses a PyTorch optimizer to apply gradients to non-embedding parameters.
+
+
+## Code
```

Comment:
Can we simplify the code snippet using pseudocode？

Reply:
Good idea. The documentation will be clearer in the description with a pseudocode. The complete demo code is for reference.